jt_notes.md

```diff
@@ -0,0 +1,127 @@
+
+**For chas:** load_data() - "Place holder function for loading data in emulator". What is the intention of this, and is there a clear idea of what an emulator would look like?
+
+**For me:** this is a general question for Chas: how does he intend the emulator to work? I don't think I've fully followed his emulator stuff in "app". Might need to talk through that one together...
+
+**For chas: IMPORTANT, BUG:** as per my comment in this module, I don't think barrier frames are implemented properly.
+
+**For chas: IMPORTANT, BUG:** I think there is a bug associated with (lack of) numExtraRefFrames in initialise_internal_parameters(), as per my comment there.
+
+**For chas: IMPORTANT, approaching a BUG**: there is the general issue that the code is not doing a great job of precise timing. It determines a delay time before sending the trigger, but then executes a bunch more code. Oh, and that delay time is then treated relative to "current_time", which is set *after* doing the phase-matching. This is going to reduce accuracy and precision, and also makes me even more uncomfortable in terms of future-proofing. I think it would be much better to pass around absolute times, not deltas.
+
+**For chas**: more generally, I wish we were working with frame *objects* that have accompanying metadata (such as timestamp!). That would help solve the above issue about timings in a tidy manner, and also make it a lot easier to generalise to other timebases (such as camera timebase, rather than computer). Oh, and it would also help deal with gaps in the timebase (missing frames), which I think would be a useful thing to cope with, even if right now there might not be any concrete scenario in which frames would be dropped. I fear I may not win that argument about frame objects, since it would be quite a big structural change. But, do you have any alternative ideas about how best to cope with the fact that we may well want to work in terms of timebases other than the computer's own timebase?
+
+Remark **for chas:** currently when we schedule a trigger everything hangs, waiting for the appropriate time to arrive. This stalls camera and sync analysis. Don't you think we should really figure out a way of doing this asynchronously, without blocking (or come to a definite conclusion that blocking is the only reliable enough way to do it on the Pi).
+
+**For me: where do the stages come in?** cli.py sets up the usb/serial link, but where is it used, by whom, and when? Should be clearer after a refactor. Either we need to conditionally interact with the stages, or we do it on a more distributed basis, in response to events. That's more how I do it in my code, and I like that approach.
```

Comment:
The stages have slowly been coming out of this codebase. And I think they should. I think the mentality of this project should be to provide a trigger, either as a high-resolution timestamp or as a timed triggger signal (Pi pins, serial object, etc., etc.) and that handling of the fluorescence camera settings, stage control and so on should be done by the user, either hardware-wise or via, e.g. Andor Solis or Nikon Elements or whatever. What do you think?

Reply:
Now I read this back to myself, I'm not sure quite what I was smoking when I wrote this. While there's a grain of sense in there somewhere, "Either we need to conditionally interact with the stages, or we do it on a more distributed basis, in response to events" is pretty heavy on buzzwords and doesn't make much sense. But I think we're in agreement here. In my later terminology, I definitely think the Sync Analyzer should have no business knowing anything about stages, hardware triggering etc. The Sync Controller _could_ (at least as an easy short-term refactor) deal with that on the RPi, but ideally in the longer term it would probably just communicate with other objects that wrap the specific hardware types.

Reply:
Stages now removed from this codebase @jmtayloruk